build_runner_core/lib/src/package_graph/build_triggers.dart

```diff
@@ -0,0 +1,201 @@
+    final buildTriggers = <String, Set<BuildTrigger>>{};
+    final warningsByPackage = <String, List<String>>{};
+    for (final entry in configByPackage.entries) {
+      final packageName = entry.key;
+      final config = entry.value;
+      final triggersByBuilder = config.triggersByBuilder;
+      for (final entry in triggersByBuilder.entries) {
+        final builderName = entry.key;
+
+        if (!builderName.contains(_builderNamePattern)) {
+          (warningsByPackage[packageName] ??= []).add(
+            'Invalid builder name: `$builderName`',
+          );
+        }
+
+        final (triggers, warnings) = BuildTriggers.parseList(
+          triggers: entry.value,
+        );
+        if (triggers.isNotEmpty) {
+          (buildTriggers[builderName] ??= {}).addAll(triggers);
+        }
+        if (warnings.isNotEmpty) {
+          (warningsByPackage[packageName] ??= []).addAll(warnings);
+        }
+      }
+    }
+    return BuildTriggers(
+      triggers: BuiltMap.from(
+        buildTriggers.map((k, v) => MapEntry(k, v.build())),
+      ),
+      warningsByPackage: BuiltMap.from(
+        warningsByPackage.map((k, v) => MapEntry(k, v.build())),
+      ),
+    );
+  }
+
+  /// Parses [triggers], or throws a descriptive exception on failure.
+  ///
+  /// [triggers] is an `Object` read directly from yaml, to be valid it should
+  /// be a list of strings that parse with [BuildTrigger.tryParse].
+  static (Iterable<BuildTrigger>, List<String>) parseList({
+    required Object triggers,
+  }) {
+    if (triggers is! List<Object?>) {
+      return (
+        [],
+        ['Invalid `triggers`, should be a list of triggers: $triggers'],
+      );
+    }
+    final result = <BuildTrigger>[];
+    final warnings = <String>[];
+    for (final triggerString in triggers) {
+      BuildTrigger? trigger;
+      String? warning;
+      if (triggerString is String) {
+        (trigger, warning) = BuildTrigger.tryParse(triggerString);
+      }
+      if (trigger != null) {
+        result.add(trigger);
+      }
+      if (warning != null) {
+        warnings.add(warning);
+      }
+    }
+    return (result, warnings);
+  }
+
+  String get renderWarnings {
+    final result = StringBuffer();
+    for (final package in warningsByPackage.keys) {
+      result.writeln('build.yaml of package:$package:');
+      for (final warning in warningsByPackage[package]!) {
+        result.writeln('  $warning');
+      }
+    }
+    result.writeln(
+      'See https://pub.dev/packages/build_config#triggers for valid usage.',
+    );
+    return result.toString();
+  }
+}
+
+/// A build trigger: a heuristic that possibly skips running a build step based
+/// on the primary input source.
+abstract class BuildTrigger {
+  /// Parses a trigger, returning `null` on failure, optionally with a warning
+  /// message.
+  ///
+  /// The only supported trigger is [ImportBuildTrigger].
+  static (BuildTrigger?, String?) tryParse(String trigger) {
```

Comment:
Could this be private?
If it has to be public I think I'd prefer a "proper return type" rather than a record.